chore(github): Update the pull request template to match new definition of done on PR

benoitf opened the pull request:
<!-- Please review the following before submitting a PR:
Che's Contributing Guide: https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
Pull Request Policy: https://github.com/eclipse/che/wiki/Development-Workflow#pull-requests
-->

### What does this PR do?
Update the pull request template to match with the new `definition of done` on pull requests.


### Screenshot/screencast of this PR
<!-- Please include a screenshot or a screencast explaining what is doing this PR -->
N/A

### What issues does this PR fix or reference?
<!-- Please include any related issue from eclipse che repository (or from another issue tracker).
     Include link to other pull requests like documentation PR from [the docs repo](https://github.com/eclipse/che-docs)
-->
Fixes https://github.com/eclipse/che/issues/17486



### How to test this PR?
you can try by opening PR against the repository https://github.com/benoitf/che-template-issues

I'm also using the PR template for this PR so you can see it live :-)


### PR Checklist

As the author of this Pull Request I made sure that:

- [x] The Eclipse Contributor Agreement is valid
- [x] Code produced is complete
- [x] Code builds without errors
- [x] Add new automated test for every bug fixed
- [x] The repository devfile is up to date and works
- [x] Sections `What issues does this PR fix or reference` and `How to test this PR` completed
- [x] Relevant user documentation updated
- [x] Relevant contributing documentation updated
- [x] CI/CD changes implemented, documented and communicated

### Reviewers

Reviewers, please add how you tested the PR when approving it.

Change-Id: I34dd7dead00e75dafd905b632bd24b81a9be24f1
Signed-off-by: Florent Benoit <fbenoit@redhat.com>

benoitf commented:
[crw-ci-test]

mmorhun commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
+
+
 ### What issues does this PR fix or reference?
+<!-- Please include any related issue from eclipse che repository (or from another issue tracker).
```
I think it would be better to have a list of issues which this PR fixes and a list of other PRs / issues on which this PR depends. Subitems works fine in my opinion.

benoitf replied:
Hello Mykola, thanks for the feedback,

I think 90% of PRs have only one link. It may be harder for new contributors to know in which section you need to just add links ?

mmorhun commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
 
-#### Release Notes
-<!-- markdown to be included in marketing announcement - N/A for bugs -->
+- [ ] The Eclipse Contributor Agreement is valid
+- [ ] Code produced is complete
+- [ ] Code builds without errors
+- [ ] Add new automated test for every bug fixed
+- [ ] The repository devfile is up to date and works
+- [ ] Sections `What issues does this PR fix or reference` and `How to test this PR` completed
+- [ ] Relevant user documentation updated
+- [ ] Relevant contributing documentation updated
+- [ ] CI/CD changes implemented, documented and communicated
```
Do we really need checkboxes here? GitHub shows a progress bar if description has them. So, the progress bar will be always shown on each PR which looks weird for me... Maybe we should go with just a list?

benoitf replied:
I think the idea was to really emphasis that items need to be done before merging it

so it may be expected to see this if all items are done
![image](https://user-images.githubusercontent.com/436777/90408628-5e555680-e0a8-11ea-8ba5-96c511898d89.png)

and this if you didn't complete all the items:
![image](https://user-images.githubusercontent.com/436777/90408834-b2603b00-e0a8-11ea-9a93-3ce8c5f69dd8.png)

sparkoo commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
 
-#### Release Notes
-<!-- markdown to be included in marketing announcement - N/A for bugs -->
+- [ ] The Eclipse Contributor Agreement is valid
```
doesn't eclipse bot do this automatically?

benoitf replied:
the main issue is that bot will do something, but basically newcomers create PR and usually forgot the SignedOff in the commits, etc.

So it's a friendly reminder before the check can happen.

sparkoo replied:
yes, but in the time of creating PR, the commits are already pushed so it's too late anyway, same as with the bot. And newcomers PRs are very rare so I don't thinkg it's worth to bother 99% PR authors for remaining 1%.

l0rd replied:
@sparkoo has a good point here. We may want to have a section of the form for newcomers with instruction to follows rather then a checkbox.

benoitf replied:
description added

sparkoo commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
-#### Release Notes
-<!-- markdown to be included in marketing announcement - N/A for bugs -->
+- [ ] The Eclipse Contributor Agreement is valid
+- [ ] Code produced is complete
```
what does it mean that "code is complete"? Is it defined somewhere?

benoitf replied:
not in work in progress anymore

vitaliy-guliy replied:
It's good to have an explanation section here, something like we have for 'How to test this PR?'.

sparkoo replied:
ok, so it's something the author decides when the PR is complete. We already have WIP bot and Draft PR for unfinished state. But I think it's still good to have it here.

Maybe rename it or add little explanation, sth like `Code complete (don't plan to do any changes except review fixes)`.

l0rd replied:
I think that applies to all the check points: we should provide a detailed and clear description and provide a link to it. I will draft something in the wiki.

benoitf replied:
description has been added

sparkoo commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
+- [ ] Code produced is complete
+- [ ] Code builds without errors
+- [ ] Add new automated test for every bug fixed
+- [ ] The repository devfile is up to date and works
```
what is "The repository devfile"?

benoitf replied:
the devfile stored in this repository to be able to use dogfooding ?

vitaliy-guliy replied:
I think it's about the devfile, which is in the repository root. It should be aligned to changes in PR.

sparkoo replied:
got it. Most of the PRs won't update that devfile, but good to remind author. +1

sparkoo commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
+- [ ] Add new automated test for every bug fixed
+- [ ] The repository devfile is up to date and works
+- [ ] Sections `What issues does this PR fix or reference` and `How to test this PR` completed
+- [ ] Relevant user documentation updated
```
I think that vast majority of PRs will check only one of these checkboxes. Can we merge them to one `Relevant documentation updated` or `Relevant user/contributing documentation updated` ?

benoitf replied:
> I think that vast majority of PRs will check only one of these checkboxes.

hello, why you'll have only one checkbox checked about documentation ? 
either you've updated everything and then it can be merged or it can't be merged because both documentation is not yet updated ?

sparkoo replied:
my point is, that most of the PRs simply does not need update both documentations and most bugfixes won't update neither of them.

l0rd replied:
@sparkoo end-user documentation and contributing documention are distinct artifacsd that target different personas. Both are important and I am afraid that if we have one unique checkbox for "documentation" we will think about end-user and forget about contributing.

sparkoo commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
-#### Docs PR
-<!-- Please add a matching PR to [the docs repo](https://github.com/eclipse/che-docs) and link that PR to this issue.
-Both will be merged at the same time. -->
+Reviewers, please add how you tested the PR when approving it.
```
does it mean that every reviewer should edit the PR description? or only note to leave a comment?

benoitf replied:
idea was to comment so maybe replace `add` by `comment`

vitaliy-guliy commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
+- [ ] Code builds without errors
+- [ ] Add new automated test for every bug fixed
+- [ ] The repository devfile is up to date and works
+- [ ] Sections `What issues does this PR fix or reference` and `How to test this PR` completed
```
It's unclear for me why do we need this, as we already have templates for both sections.

benoitf replied:
it's just to not let these sections 'blank'

sparkoo replied:
isn't that already explicit ehough? It's like add another checkbox `[ ] all checkboxes above are checked`

benoitf replied:
yes but a lot of PR have no description or how to be able to test it paragraph even if it sounds logical to explain how to describe what the PR is doing and how to validate it

Also on some PR, the template is totally removed so it's to ensure consistency

metlos commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
-<!-- markdown to be included in marketing announcement - N/A for bugs -->
+- [ ] The Eclipse Contributor Agreement is valid
+- [ ] Code produced is complete
+- [ ] Code builds without errors
```
Don't we have a CI build that will prevent merging the PR if the build fails?

l0rd replied:
Correct. That means that contributors should verify that code builds successfully before opening a PR.

themr0c replied:
The build will fail if Antora fails, but not if other validations fail: linkchecker issues, linter (vale) errors, etc.

metlos commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,24 +1,43 @@
+- [ ] The Eclipse Contributor Agreement is valid
+- [ ] Code produced is complete
+- [ ] Code builds without errors
+- [ ] Add new automated test for every bug fixed
```
The imperative form seems doesn't seem appropriate for a checklist. What about something like "The bugfix/feature is covered by tests."?

benoitf replied:
in fact, when checking with vale plug-in (the one used in documentation) it says that we should not use 'passive form'

l0rd replied:
@themr0c do you know how we can make vale happy?

themr0c replied:
What about: 
"Tests are covering the bugfix or new feature"

l0rd commented:
@benoitf I have submitted https://github.com/eclipse/che/pull/17834. As a side note I have changed "Tests are covering the bugfix or new feature" with "Tests are covering the bugfix". That's because I don't want to enforce the creation of new tests for every PR but only if it fixes a bug.

benoitf commented:
@l0rd PR has been updated

I've added links to each element to `CONTRIBUTING.md` file

I've also updated the description of this PR to match the content of the template to see how it looks live

l0rd commented:
@benoitf LGTM
